File: 2020_leetcode/problems/__01__/kotlin/leetcode/RecursionPrinciple.kt
package leetcode

object RecursionPrincipleKt {
    @JvmStatic fun printReverse(str: CharArray) {
        fun helper(index: Int, str: CharArray) {
            if (index >= str.size) {
                return
            }
            helper(index + 1, str)
            println(str[index])
        }
        helper(0, str)
    }
}

File: 2020_leetcode/problems/__01__/rust/recursion_principle.rs
fn print_reverse(str: &str) {
    let char_vec = String::from(str).chars().collect::<Vec<char>>();
    fn helper(index: usize, char_vec: &Vec<char>) {
        if index >= char_vec.len() {
            return;
        }
        helper(index.clone() + 1, &char_vec);
        println!("{:?}", *char_vec.get(index).unwrap());
    }
    helper(0, &char_vec);
}

fn main() {
    print_reverse("abc");
}
